Add engine.softTakeoverIgnoreNext() to fix bug #1520798.

Pegasus-RPG opened the pull request:
This adds the ability for scripts to call SoftTakeover::IgnoreNext() to solve the problem presented in bug #1520798 where an absolute control is remapped on-the-fly in script between soft-takeover protected COs. Changed the VMS4 script to use this capability and it tests good. The soft-takeover ignore algorithm itself was not changed (just some spelling corrections in comments.)

Pegasus-RPG committed:
Added ability for scripts to tell soft-takeover to ignore the next value, providing a solution to bug #1520798

Pegasus-RPG committed:
Removed inaccurate comment in VMS4 script

Pegasus-RPG committed:
Removed inaccurate comment in VMS4 script, in another spot

daschuer commented on src/controllers/softtakeover.h:
```diff
@@ -13,13 +13,11 @@
-    static const double kDefaultTakeoverThreshold;
+    // 3/128 units away from the current is enough to catch fast non-sequential moves
+    //  but not cause an audibly noticeable jump.
+    static const double kDefaultTakeoverThreshold = 3.0 / 128;
```
Did you test this with non c++11 settings and on all targets? I thought initializing a static variable in the class definition is not allowed.

Pegasus-RPG replied:
Crap, that didn't even cross my mind. (The joys of developing on Linux...)
I can revert that bit. I don't have the ability to test everywhere right now.

Pegasus-RPG replied:
Wait a sec, it's done with "static const uint SUBSEQUENT_VALUE_OVERRIDE_TIME_MILLIS = 50;" so if that works, this should too, right?

daschuer replied:
Sorry for the false positive. I have just tested it on Ubuntu Trusty and surprisingly to me it compiles. 
The same on Travis.

ywwg replied:
buildbot: test this please

rryan replied:
Either way it's a feature that was introduced in c++11 -- your compilers are being nice and not failing because they already have the feature implemented and you didn't specifically build with ```-std=c++03``` :)

rryan replied:
Oops no I'm wrong this is different than the c++11 feature I'm thinking of.

rryan replied:
Huh well clang actually warns:
```
In file included from src/controllers/controllerengine.h:18:
src/controllers/softtakeover.h:20:25: warning: in-class initializer for static data member of type
      'const double' is a GNU extension [-Wgnu-static-float-init]
    static const double kDefaultTakeoverThreshold = 3.0 / 128;
                        ^                           ~~~~~~~~~
```

Pegasus-RPG replied:
okay so I'll move both into the CPP file.

daschuer commented on src/controllers/controllerengine.cpp:
```diff
@@ -1440,6 +1440,22 @@ void ControllerEngine::softTakeover(QString group, QString name, bool set) {
+     Input:   ControlObject group and key values
+     Output:  -
+     -------- ------------------------------------------------------ */
+void ControllerEngine::softTakeoverIgnoreNext(QString group, QString name) {
```
softTakeoverIgnoreNext() might be a puzzel if you do not know the Implementation details. 

This function must we called whenever a absolute control takes over a ControlObject which was controlled from elsewhere and which is likely out of sync because of that. 
It would be nice if we can relfect something of this in the function name.

softTakeoverRequireResync()
softTakeoverDetach() 
softTakoverLoose()  
softTakoverAtacheOtherControl()

Pegasus-RPG replied:
What if I just call it softTakeoverSync?

daschuer replied:
I would prefer Detach over Resync since the control is not in sync after the call. 
But I do not want to stress this any further, jut pick your favorite.

rryan committed:
Add X-Session pro wiki link.

Pegasus-RPG committed:
Renamed function exposed to script to softTakeoverResync to make it a bit clearer when it should be used

esbrandt committed:
Revert b3841384b7a817e1c9bc9e3ae816487820099efd
Reverting this commit fixes that custom widgets (e.g. Preview column, Cover column for tracks with no cover) do not correctly inherit the selected line color.

esbrandt committed:
Fixes tree expansions symbols are not shown in Deere skin, https://bugs.launchpad.net/mixxx/+bug/1504918.
Symbol files went MIA in previous commits

ywwg committed:
Merge pull request #797 from esbrandt/deere_treeview

Deere treeview fix

ywwg committed:
Merge pull request #798 from esbrandt/tooltip_updates

Tooltip updates

esbrandt committed:
Update translation templates. Found 2006 source text(s) (51 new and 1955 already existing). Kept 4 obsolete entries

esbrandt committed:
Deere: Add back images for sorting indicators in the library header. They were deleted in a previous commit. Fixes error message on skin load.

esbrandt committed:
Deere: Fix bug where content of <TooltipId> would not be displayed in the tooltips ( e.g. previously for the ``repeat`` control in the deck/sampler). Got fixed by removing a unused <Tooltip> variable. Looks like <Tooltip> and <TooltipId> variables can not coexist in templates.

esbrandt committed:
Deere: Add tooltips to controls in the Effect Rack widget. Update tooltips in the mixer widget.

esbrandt committed:
Deere: Add ``parameter_link_type`` button below effect parameter to control how linked effects change when turning the super knob.    not linked / linked / split linked left / split linked right  / split linked left&right . EQ Gain effects cannot be used in split mode (not linked / linked only). Follow up to #321

rryan committed:
Update manual build to cd1b20bd76b7d854b6f451aa97e3a01bcfe6c588.

esbrandt committed:
Deere: Simplify hamburger menu button style in the skin settings widget. Button inherits now the skins default button style. Follow up to d5c0844e11588755f93c2c68b1a321305916a560

esbrandt committed:
Deere: Stylesheet fixes
* Update scrollbar style to match existing design language. Add hover style.
* Simplify some overqualified element names

rryan committed:
Add protection against null NetworkStreamWorker in EngineNetworkStream. Fixes segfault on startup when built with broadcasting disabled.

rryan committed:
Bundle mixxx.pdb and mixxx-test.pdb.

rryan committed:
Argh.

rryan committed:
Add 1.10 and 1.11 OS X SDK versions.

daschuer committed:
invert isRunnig test in EngineShoutcast

daschuer committed:
Improve error message in ~EngineShoutcast()

ywwg committed:
controllers: clean up stray whitespace in common-hid

ywwg commented:
I am running in to the same issue with the traktor mapping, so let's get this in

ywwg commented on src/controllers/controllerengine.h:
```diff
@@ -107,6 +107,7 @@ class ControllerEngine : public QObject {
     Q_INVOKABLE void scratchDisable(int deck, bool ramp = true);
     Q_INVOKABLE bool isScratching(int deck);
     Q_INVOKABLE void softTakeover(QString group, QString name, bool set);
+    Q_INVOKABLE void softTakeoverResync(QString group, QString name);
```
can you call it reset instead of resync?  It could be interpreted to mean that the softtakeover is ignored

Pegasus-RPG replied:
It /is/ ignored on the next iteration.

ywwg replied:
I mean ignored as in the control is warped to the new value.  like "oh resync, that's a way to override soft takeover".

reset makes it a little clearer I think

rryan replied:
haha we're both shaving this yak in different places

daschuer replied:
Reset is in the same way confusing as Resync since the scope is fuzzy. Lets stick with Resync or long name that describes exactly what it does or when it should be called. 
Something like: 

softTakeoverFromAnOutOfSyncPot()

ywwg commented on src/controllers/softtakeover.cpp:
```diff
@@ -94,11 +101,11 @@ bool SoftTakeover::ignore(ControlObject* control, double newParameter) {
         const double prevDiff = currentParameter - m_prevParameter;
         if ((prevDiff < 0 && difference < 0) ||
```
Even though the logic for ==0 is never tested, because diffs at zero are never greater than the threshold I think this logic is actually ok

rryan commented:
Probably worth adding unit tests to guard against future breakage.

ywwg commented:
what's our target date for RC2, and @Pegasus-RPG do you think you can figure out how to write a test by then?  Or we can write one for you if that's a big lift.  This is a blocker for the S4 mapping so time is a factor

rryan commented on src/controllers/controllerengine.cpp:
```diff
@@ -1440,6 +1440,23 @@ void ControllerEngine::softTakeover(QString group, QString name, bool set) {
+     Input:   ControlObject group and key values
+     Output:  -
+     -------- ------------------------------------------------------ */
+void ControllerEngine::softTakeoverResync(QString group, QString name) {
```
Why "resync" if the code and description say "ignoreNext" ?

Pegasus-RPG replied:
Because it allows one to re-synchronize a control with SoftTakeover instead of it just jumping which is the bug.

Pegasus-RPG replied:
(I'm fine with calling it ignoreNext() FTR.)

ywwg replied:
SGTM

rryan replied:
I'd just prefer everything to be called one thing -- maybe rename ignoreNext on the SoftTakeover side then ?

Pegasus-RPG commented:
If you can at least write the shell, I can add the logic. (I still don't understand how and which objects need to be instantiated for tests.)

rryan commented:
> If you can at least write the shell, I can add the logic. (I still don't understand how and which objects need to be instantiated for tests.)

Sure -- I'll send you a PR

ywwg commented:
We already have controllerengine_test.cpp so you can use that --- and RJ just posted :P

rryan committed:
Switch Time::elapsedMsecs to return qint64.

When dealing with elapsed times you almost always want signed integers
because zero crossings can cause crazy large numbers.

For unsigned integers: (small number - small number) => GIGANTIC NUMBER
For signed integers: (small number - small number) = +/- small number

rryan committed:
Switch Time::elapsedMsecs to return qint64.

When dealing with elapsed times you almost always want signed integers
because zero crossings can cause crazy large numbers.

For unsigned integers: (small number - small number) => GIGANTIC NUMBER
For signed integers: (small number - small number) = +/- small number

rryan committed:
Add "test mode" to Time that allows test to control how much time has elapsed.

rryan committed:
Add "test mode" to Time that allows test to control how much time has elapsed.

rryan committed:
Add SoftTakeover test fixture and a few basic tests.

Pegasus-RPG committed:
Merge pull request #1 from rryan/st_fix

Add a basic test fixture for SoftTakeover.

Pegasus-RPG committed:
Fixed C++03 violation, changed function name to ignoreNextValue(). This is hereby the canonical name, so if you don't like it, add aliases in ControllerEngine on your own time. :)

Pegasus-RPG committed:
Added soft-takeover tests and some Time improvements from RJ to make them work.

Pegasus-RPG commented:
Put the brakes on this review a minute. I've just pushed some ST tests to 1.12/2.0 that worked in 1.11 but some fail in 2.0, suggesting a logic error in 2.0. Working on it, would welcome extra eyes on the tests and ST::ignore() as well.

Pegasus-RPG committed:
More SoftTakeover tests to check for sign inaccuracies in the ignore cases

Pegasus-RPG committed:
Re-derived the ST ignore cases with a truth table, now in the comments in softtakeover.cpp. Adjusted the tests to match the truth table.

Pegasus-RPG committed:
More comments on ST tests, now uses time threshold from softtakeover.h for less hard-coding

Pegasus-RPG committed:
WIP: Changed the softtakeover algorithm but having problems getting the tests to work correctly.

Pegasus-RPG committed:
Clarified units in Time class, allowing me to fix problems with softTakeover tests. My new code now passes them all! (Old code failed 8.)

Pegasus-RPG commented:
This is now obsolete. The real problem is fixed in PR #804. Thanks very much to all of you for your input and guidance!

Pegasus-RPG commented:
This is back on the table. It turns out it's required so as not to break the ability to move controls quickly.

Pegasus-RPG commented:
Apparently I need to clean it up. Merge #804 first and I'll base this on that.

Pegasus-RPG committed:
- Corrected time scales in Time class (the timer is in nanoseconds!)
- Addressed PR notes
- Added tests for quick whipping of controls at the edges of the ControlObjects (when the previous value matches the current)

Pegasus-RPG committed:
Comment grammar fix

Pegasus-RPG committed:
Merged from st_real_fix

Pegasus-RPG commented:
I guess I merged something incorrectly since a bunch of stuff is showing as changed here when I only meant to include my changes.

Pegasus-RPG commented:
I jacked this up. Moving to a new PR.
